lib/puppet_x/puppetlabs/scheduled_task/trigger.rb

```diff
@@ -142,16 +143,20 @@ def self.canonicalize_and_validate(manifest_hash)
         raise ArgumentError.new("Unknown schedule type: #{manifest_hash["schedule"].inspect}")
       end
 
+      # The boot trigger does not require a start time because it is an event
+      # based trigger, not a schedule. This is handled as a one off for now
+      # until we can make the changes to the module required to handle event
+      # based triggers as a general case.
       # required fields
       %w{start_time}.each do |field|
         next if manifest_hash.key?(field)
-        raise ArgumentError.new("Must specify '#{field}' when defining a trigger")
+        raise ArgumentError.new("Must specify '#{field}' when defining a trigger") unless manifest_hash['schedule'] == 'boot'
```

Comment:
I agree that checking for this inside a loop isn't great. In my previous implementation I had checked for the general case of `EVENT_BASED_TRIGGER.value?(manifest_hash['schedule'])`. That check was done in an unless block that prevented this code from running at all if it was an event trigger. That way this would never have to be modified until we ran into more complex requirements than either or, in which case this would likely need to be moved into a function anyway. 

If you don't think that would be better I'll go ahead and make this modification.

Reply:
Yeah, I think the condition is kind of hiding in the loop, which makes for a little bit of a trap if something else is added to the required list with `start_time` (at the very least it would make for multiple code changes instead of just one)... 

The check could also be moved up a line higher to say `next if manifest_hash.key?(field) || (manifest_hash['schedule'] == 'boot')` but I think that suffers the same problem as what we've got now.

I think setting the required fields up-front is clearer in this case.

Reply:
To clarify though what I meant was more like 
```
%w{start_time}.each do |field|
  next if manifest_hash.key?(field)
  raise ArgumentError.new("Must specify '#{field}' when defining a trigger")
end unless EVENT_TRIGGER_COLLECTION.value?(manifest_hash['schedule'])
# Is it bad form to put an unless at the end of a block like this?
```
This way the entire loop is skipped if it's a boot trigger, you can add fields to the array as needed, and even if we support new event based triggers in the future we still don't need to edit this check. Adding the full list of event based triggers to an event based triggers collection doesn't make it a free for all for people to put those trigger types in the manifest because we just wouldn't add types we haven't tested to the `ValidScheduleKeys` array.

So yeah, if that kind of check is also a bad idea I'll implement your original suggestion.

Reply:
Yeah, you can do the conditional after the `end` like that, but it's usually considered bad form for readability reasons - see Rubocop rule at https://github.com/bbatsov/ruby-style-guide#no-multiline-if-modifiers

I think we'll probably end up with something like your idea eventually: 

```ruby
required = EVENT_TRIGGER_COLLECTION.value?(manifest_hash['schedule']) ? [] : %w{start_time}
```

But hopefully it's based off the model idea I was tossing around and is more like:

```ruby
required = TRIGGER_DEFINITIONS[manifest_hash['schedule']].required
```

Reply:
Well, until we get the model definitions idea worked out, we could move the unless from the end of the block to an enclosing
```
unless <condition>
  <do stuff>
end
```

Reply:
An empty array `[]` for required fields or an `unless` like that are both fine I think... either way.